File: WordBreak/word_break.py
#!/usr/bin/env python3
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = [False]*(len(s)+1)
        memo[len(s)] = True

        for i in range(len(s)-1, -1, -1):
            for word in wordDict:
                if (i+len(word) <= len(s)
                    and word == s[i: i+len(word)]):
                    memo[i] = memo[i+len(word)]
                if memo[i]:
                    break
        return memo[0]
                    
            
        

File: WordBreak/word_break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<int> dp(s.size()+1, false);
        dp[0] = true;
        for (int i = 1; i < s.size()+1; ++i){
            for (auto& word : wordDict){
                if (word.size() <= i && s.substr(i-word.size(), word.size()) == word){
                    dp[i] = dp[i] || dp[i - word.size()];
                if (dp[i]) 
                    break;
                }
            }
        }
        return dp[dp.size()-1];
    }
};

